src/distribution/cauchy.rs

```diff
@@ -0,0 +1,238 @@
+pub struct Cauchy {
+    loc: f64,
+    gamma: f64,
+}
+
+impl Cauchy {
+    /// Create a Cauchy distribution with location `loc` and scale `gamma`.
+    ///
+    /// It should hold that `gamma > 0`.
+    #[inline]
+    pub fn new(loc: f64, gamma: f64) -> Self {
+        should!(gamma > 0.0);
+        Cauchy { loc, gamma }
+    }
+
+    // Return the location parameter
+    #[inline(always)]
+    pub fn loc(&self) -> f64 {
+        self.loc
+    }
+
+    // Return the scale parameter
+    #[inline(always)]
+    pub fn gamma(&self) -> f64 {
+        self.gamma
+    }
+}
+
+impl distribution::Continuous for Cauchy {
+    #[inline]
+    fn density(&self, x: f64) -> f64 {
+        // Divisions are expensive, so we implement the PDF with only a single division.
+        use std::f64::consts::PI;
+        let deviation = x - self.loc;
+        self.gamma / (PI * (self.gamma * self.gamma + deviation * deviation))
+    }
+}
+
+impl distribution::Distribution for Cauchy {
+    type Value = f64;
+
+    #[inline]
+    fn distribution(&self, x: f64) -> f64 {
+        use std::f64::consts::FRAC_1_PI;
+        FRAC_1_PI * ((x - self.loc) / self.gamma).atan() + 0.5
+    }
+}
+
+impl distribution::Entropy for Cauchy {
+    #[inline]
+    fn entropy(&self) -> f64 {
+        (std::f64::consts::PI * 4.0 * self.gamma).ln()
+    }
+}
+
+impl distribution::Inverse for Cauchy {
+    /// Due to finite precision of arithmetic operations, the current implementation of
+    /// `Inverse::inverse` for `Cauchy` does *not* return negative or positive infinity for
+    /// `p = 0.0` or `p = 1.0`, respectively. It instead returns very large (in magnitude)
+    /// values (`> 1e16`).
+    #[inline]
+    fn inverse(&self, p: f64) -> f64 {
+        should!((0.0..=1.0).contains(&p));
+        use std::f64::consts::PI;
+        self.loc + self.gamma * (PI * (p - 0.5)).tan()
+    }
+}
+
+impl distribution::Median for Cauchy {
+    #[inline]
+    fn median(&self) -> f64 {
+        self.loc
+    }
+}
+
+impl distribution::Modes for Cauchy {
+    #[inline]
+    fn modes(&self) -> Vec<f64> {
+        vec![self.loc]
+    }
+}
+
+impl distribution::Sample for Cauchy {
+    #[inline]
+    fn sample<S>(&self, source: &mut S) -> f64
+    where
+        S: Source,
+    {
+        // We use the fact that the ratio of two standard normal random variables is
+        // standard Cauchy distributed. This way of drawing a sample from a Cauchy
+        // distribution turned out to be more efficient in the benchmarks than the naive way
+        // of applying the quantile function to a uniformly distributed random variable.
+        // However, the employed method requires reading more random numbers from the
+        // source, so it might be slower for slow random sources.
+        let gaussian = distribution::Gaussian::new(0.0, 1.0);
+        let a = gaussian.sample(source);
+        let b = gaussian.sample(source);
+        self.loc() + self.gamma() * a / (b.abs() + f64::EPSILON)
```

Comment:
Not sure about the epsilon. Do you have a reference for this implementation?

Reply:
Oh, good catch! This should be `f64::MIN_POSITIVE`. (Adding `MIN_POSITIVE` has no effect unless `b.abs()` is smaller than `1.0e-291`, in which case we're essentially dividing by zero.)